### DISCUSSION TITLE
Readonly parameters in primary contructor

### DISCUSSION THREAD

**Discussion by Krzysztof318** [dotnet/csharplang#7699] [Language Ideas] [answered]
Hi, it will be very convenient if we can create readonly parameters in primary contructors, escpecially for injecting services. I don't want to variable (for example IService type) be mutable.

Idea example:
`public class ExampleClass(readonly IService service) { }`

**Comment by DavidArno** [comment, marked as answer]
You and about 100,000 other devs want this feature. The last we heard from the team on this is documented in [C# Language Design Meeting for July 31st, 2023](https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-07-31.md):

> ... Primary constructor parameters will ship as mutable-by-default, and we will investigate the `readonly` parameter modifier shortly...

Worryingly, in #7667, MS were still asking the question, "*What else would you like to see and what are the important scenarios to you? ... Support for immutability in primary constructor parameters?*" just last week. Whether that implies that either we haven't arrived at that "shortly" period of time yet, or the team are backtracking on ever doing this is unclear at the moment.

**Reply by HaloFour** [reply]
I think that it's good to elicit feedback on the subject, particularly since those discussions get linked through blog posts with a wider circulation so hopefully it would attract folks who don't normally lurk here.

**Reply by CyrusNajmabadi** [reply]
> Worryingly

Nothing has changed with our plans here. Indeed, we just talked about getting onto this, now that everything has settled.

**Reply by Rowe2ryWA** [reply]
Chiming in here in September 2025, a few .Net releases and C# versions later... we're still not able to do this. Would love to be able to decorate primary constructor params with something like:

```
public class MyClass([AsReadonly]ISomeSingletonServiceOrsomething service)
{
}
```

**Reply by TahirAhmadov** [reply]
@Rowe2ryWA 
https://github.com/dotnet/csharplang/discussions/8715#discussioncomment-11453606

**Comment by Krzysztof318** [comment]
I'm sorry. I don't know how to was possible I didn't see this.

**Reply by HaloFour** [reply]
It's not possible today, but it's something that the team has expressed interest in exploring based on feedback.

**Reply by Krzysztof318** [reply]
I meant that was my fault that I didn't see this case is know already.

**Comment by Krzysztof318** [comment]
The idea that came to my mind is that since the primary constructor has parameters similar to regular constructors or methods, it would be a good idea to add the ability to set "readonly" in parameters for both methods and constructors. Parameters are usually used only for reading data, and they are rarely changed, so it would be beneficial to introduce optional protection against accidental parameter overwriting. By the way, do I need to create a new thread for this extended idea to be noticed by someone from Microsoft's team?

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/issues/188
https://github.com/dotnet/csharplang/blob/main/proposals/readonly-parameters.md
